### DISCUSSION TITLE
Connections remain after db.close() on 3.2.10

### DISCUSSION THREAD

**Discussion by makandre** [orientechnologies/orientdb#9879] [Q&A] [answered]
I'm seeing a connections leak and I can't figure out what I'm doing wrong.

My setup is this:  I'm running the orientdb:3.2.10 image from dockerhub.  My application is a spring boot application connecting to the container, using `remote:localhost` (on default port 2424)

And I'm not doing anything fancy.  From what I can see, the simple step of opening a database (`orientdb.open(database, user, password)`) creates 4 connections as seen in servers management dashboard

![image](https://user-images.githubusercontent.com/17596240/195718020-559f4a86-3a58-4094-9384-fe3adb1934d1.png)

Then, without doing anything I just call `db.close()` and only 1 of those connections disappear, the other 3 never goes away. It seems doing other db operations just piles on more connections, and `db.close()` does not appear to do anything.  As you can imagine, the connections keep increasing and then I get this error

```
WARNI Reached maximum number of concurrent connections (max=1000, current=1705), reject incoming connection from /10.89.0.26:47470 [OServerNetworkListener]
```
 
Any help?  Thanks in advance.

**Comment by makandre** [comment]
P.S. I have tried switching to using a connection pool, but it doesn't help and the same connections leak is still observed.

**Comment by tglman** [comment, marked as answer]
Hi,

The connections will be terminated when the OrientDB context will be closed, not when the session result of the open is closed.

Regards

**Reply by makandre** [reply]
Then I don't quite understand how connection pooling can be used, if we are supposed to close the OrientDB context, and create a new one each time we want to open a db connection?  Could you provide some sample code?

**Reply by tglman** [reply]
Hi,

OrientDB does internal multiplexing of connections (tcp/ip sockets), the `ODatabasePool` is a pool of sessions that allow you to optimize session handshaking and limit the number of concurrent executions, there is not strict relations between tcp/ip connections and sessions, some connections will be kept open for the whole life of the OrientDB context to handle client side metadata refresh.

just to clarify:  a session is the result of an open operation.
there is no way for you to manually control the number of tcp/ip connections the OrientDB context will handle it for you with some configurable limits (99% of time default limits are good).

**Reply by makandre** [reply]
Ok.. I understand it a little better, but please bear with me for a minute.  Let's say this is what I do:

1. create context
2. create session with open
3. perform many db operations <-- more connections are created with each operation
4. close session
5. close context

At step 3, it appears that a large number of connections will get opened, possibly hitting the connections limit.  So is what I observe accurate and expected then?  Should I actually create a new context/session for each db operation instead?

A slightly related question, how would I increase the connections limit if I had to?

Appreciate the responses.

**Reply by makandre** [reply]
Just to give some additional context about what step 3 might be doing, an example is I'm creating the schemas in a newly created database.  So I'm making a number of `createVertexClass()`/`createProperty()`/`createIndex()/etc` calls, and each of them results in new connections that can be seen in the server management dashboard.  It would not seem particular performant if I had to create and close a context and session for each of those calls.   Please advise.

**Reply by tglman** [reply]
For what you are saying sounds like you are using different versions between client and server, at your step 3 if you have only one session and one thread, I do not expect that are created more then 1 or 2 tcp/ip connections.

It does happen that if you have different version in the client and the server may be there some network failure that cause the connection number to grow.

**Reply by makandre** [reply]
No I'm using same versions.  3.2.10 server and client.  When I temporarily reverted to 3.2.4 to test, it was also reverting both server and client.

**Reply by makandre** [reply]
This seems a bit crazy, I'm stepping through my code, and I find that

```
Iterable<OEdge> edges = vertex.getEdges(ODirection.OUT);
for (OEdge edge : edges) {
    // do something
}
```

Even just iterating here, each iteration is creating a new connection.  Is it somehow lazily connecting to fetch an edge each time?

**Reply by makandre** [reply]
EUREKA!  My problem is TIME!

As it turns out, it's because I'm running the server in a container--using Podman to be precise, and Podman machine's system time has significantly drifted (apparently a known problem on mac).  This is what's causing each operation to result in a new connection because the time of the server and client were not in sync!

@tglman thank you for your replies in helping me track down the cause.

**Reply by tglman** [reply]
Hi,

Good to know you could trace back the problem, and to let us know it here, this will help anyone that come trough the same problem in future.

**Comment by makandre** [comment]
@tglman Also, right before you replied I actually opened an issue out of this discussion, because what I noticed is that on version 3.2.4, the behaviour is different.  It's slightly better: only 2 connections is left behind instead of 3.

And the problem goes beyond just opening/closing the db (that was for illustrating the problem).  There's still the fact that performing db operations (such as queries), appears to create **NEW** connections too, so in a complex application doing many operations, it's quite possible to hit the connections limit.
